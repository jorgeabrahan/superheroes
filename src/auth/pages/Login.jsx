import { useContext } from "react";
import { useNavigate } from "react-router-dom";
import { AuthContext } from "../context/AuthContext";

export const Login = () => {
  const navigate = useNavigate()
  const { login } = useContext(AuthContext)

  const handleLogin = () => {
    login('Jorge')
    const lastPath = localStorage.getItem('lastPath') || '/'
    navigate(lastPath, { replace: true });
  }
  return (
    <>
      <div className="container mt-5">
        <h1>Login</h1>
        <hr />
        <button className="btn btn-primary" onClick={handleLogin}>
          Login
        </button>
      </div>
    </>
  );
}
