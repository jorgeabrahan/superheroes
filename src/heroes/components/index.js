export * from './HeroCard'
export * from './HeroList'
export * from './Navbar'
