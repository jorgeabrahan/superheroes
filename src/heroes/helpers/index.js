export * from './getHeroesByPublisher'
export * from './getHeroById'
export * from './getHeroesByName'
