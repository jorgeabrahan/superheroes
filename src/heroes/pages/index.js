export * from './Dc'
export * from './Marvel'
export * from './Search'
export * from './Hero'
