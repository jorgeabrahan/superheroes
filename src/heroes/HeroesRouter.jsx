import { Navigate, Route, Routes } from 'react-router-dom'
import { Dc, Hero, Marvel, Search } from './pages'
import { Navbar } from './components'

export const HeroesRouter = () => {
  return (
    <>
      <Navbar />
      <div className="container">
        <Routes>
          <Route path="marvel" element={<Marvel />} />
          <Route path="dc" element={<Dc />} />

          <Route path="search" element={<Search />} />
          <Route path="hero/:id" element={<Hero />} />

          <Route path="/" element={<Navigate to="/marvel" />} />
        </Routes>
      </div>
    </>
  )
}
