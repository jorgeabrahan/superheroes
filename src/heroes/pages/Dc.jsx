import { HeroList } from '../components'

export const Dc = () => {
  return (
    <>
      <h1>Dc Comics</h1>
      <hr />

      <HeroList publisher="DC Comics" />
    </>
  )
}
