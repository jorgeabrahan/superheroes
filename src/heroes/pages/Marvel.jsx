import { HeroList } from '../components'

export const Marvel = () => {
  return (
    <>
      <h1>Marvel Comics</h1>
      <hr />

      <HeroList publisher="Marvel Comics" />
    </>
  )
}
