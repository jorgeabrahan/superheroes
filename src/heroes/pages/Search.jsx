import { useNavigate, useLocation } from 'react-router-dom'
import queryString from 'query-string'
import { useForm } from '../../hooks/useForm'
import { HeroCard } from '../components'
import { getHeroesByName } from '../helpers'

export const Search = () => {
  const navigate = useNavigate()
  const location = useLocation()
  const { q = '' } = queryString.parse(location.search)
  const heroes = getHeroesByName(q)
  const displayNoResults = q !== '' && heroes.length === 0

  const { searchQuery, onInputChange } = useForm({
    searchQuery: q
  })

  const onSearch = (e) => {
    e.preventDefault()
    navigate(`?q=${searchQuery}`)
  }

  return (
    <>
      <h1>Search</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>Searching</h4>
          <hr />
          <form onSubmit={onSearch}>
            <input
              className="form-control"
              type="text"
              name="searchQuery"
              id="searchQuery"
              placeholder="Search a hero"
              autoComplete="off"
              value={searchQuery}
              onChange={onInputChange}
            />
            <button className="btn btn-outline-primary mt-1">Search</button>
          </form>
        </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr />
          {q === '' && (
            <div className="alert alert-primary animate__animated animate__fadeIn">
              Search a hero
            </div>
          )}
          {displayNoResults && (
            <div className="alert alert-danger animate__animated animate__fadeIn">
              No results for {q}
            </div>
          )}

          {heroes.map((hero) => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </>
  )
}
