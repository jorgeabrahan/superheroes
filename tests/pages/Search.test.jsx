import { describe, test, jest, expect } from '@jest/globals'
import { fireEvent, render, screen } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import { Search } from '../../src/heroes/pages'

const mockedUseNavigate = jest.fn()

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUseNavigate
}))

describe('Pruebas en <SearchPage />', () => {
  test('Debe de mostrarse correctamente con valores por defecto', () => {
    const { container } = render(
      <MemoryRouter>
        <Search />
      </MemoryRouter>
    )
    expect(container).toMatchSnapshot()
  })

  test('Debe de mostrar a batman y el input con el valor del query string', () => {
    render(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <Search />
      </MemoryRouter>
    )
    const input = screen.getByRole('textbox')
    expect(input.value).toBe('batman')

    const img = screen.getByRole('img')
    expect(img.src).toContain('/assets/heroes/dc-batman.jpg')
  })

  test('Debe de mostrar un error si no se encuentra un hero', () => {
    render(
      <MemoryRouter initialEntries={['/search?q=batman123']}>
        <Search />
      </MemoryRouter>
    )
    expect(screen.getByText('No results for batman123')).toBeTruthy()
  })

  test('Debe de llamar el navigate a la pantalla nueva', () => {
    render(
      <MemoryRouter initialEntries={['/search']}>
        <Search />
      </MemoryRouter>
    )
    const input = screen.getByRole('textbox')
    fireEvent.change(input, { target: { name: 'searchQuery', value: 'superman' } })
    const form = screen.getByRole('form')
    fireEvent.submit(form)
    expect(mockedUseNavigate).toHaveBeenCalledWith('?q=superman')
  })
})
