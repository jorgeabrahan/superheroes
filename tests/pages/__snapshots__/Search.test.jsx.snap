// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage /> Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          id="searchQuery"
          name="searchQuery"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
